internal/bson/bson.go

```diff
@@ -59,7 +59,7 @@ import (
 type bsontype interface {
 	bsontype() // seal for go-sumtype
 
-	ReadFrom(*bufio.Reader) error
+	ReadFrom(*bufio.Reader, int) error
```

Comment:
Well, this `int` without any explanation looks a bit confusing here. Also, it doesn't feel like it really belongs to this interface. If I have a simple type, I don't have any depth.

I understand why you introduced it - you need to know what is the current depth.
But to me, it rather looks like there is one more interface here, that specifies that nested nature - only for arrays and documents. Then you can keep this `ReadFrom` as it is, and you don't need to have calls like `doc.ReadFrom(bufr, 0)`.


Something like 

```
type bsonnested interface {
    ReadFrom(*bufio.Reader, int) error
```

and only Array and Document should implement it.

And then the original ReadFrom will be the one that calls this "nested" ReadFrom. 
This way, we hide that internal information about nestedness from the initial callers of `ReadFrom`.

If my idea is not clear, just ping me for a quick pairing session.

Reply:
yeah, sounds good to me too!

Reply:
There is also no need to export those methods